qt/python/mantidqt/widgets/sliceviewer/peaksviewer/model.py

```diff
@@ -92,15 +92,12 @@ def add_peak(self, pos, frame):
 
     def delete_peak(self, pos, frame):
         r"""Delete the peak closest to the input position"""
-        frame_name = str(frame)
-        if frame_name == 'QLab':
-            positions = np.array(self.peaks_workspace.column('QLab'))
-        elif frame_name == 'QSample':
-            positions = np.array(self.peaks_workspace.column('QSample'))
-        elif frame_name == 'HKL':
-            positions = np.array([peak.getHKL() for peak in self.peaks_workspace])
-        else:
-            raise ValueError(f'Frame {frame_name} not understood')  # should not reach here
+
+        if self.peaks_workspace.getNumberPeaks() == 0:
+            return
+
+        frame_to_slice_fn = self._frame_to_slice_fn(frame)
```

Comment:
@rosswhitfield  I had noticed `_frame_to_slice_fn` but decided not to use it because I thought that self.peaks_workspace.column would be  faster and might be of some advantage when having thousands of peaks. Do you know if both approaches are equally fast?

Reply:
For 100000 peaks

```python
peaks.column('QSample')
```

takes ~0.04 seconds


```python
[peak.getQSampleFrame() for peak in peaks]
```

takes ~0.05 seconds

So yes it's about 25% slower. Do we care?

Reply:
Just in case, I tried with `[getattr(peak, 'getQSampleFrame')() for peak in peaks]` but it only adds about another 0.01s so we'll leave it as it is

Reply:
By far the slowest part of all this is converting to a numpy array, I think it's because we are converting a list of V3D into a number array.

So for 100000 peaks

```python
np.array(peaks.column('QSample'))
```

takes about the same time (about 3 seconds) as

```python
np.array([getattr(peak, 'getQSampleFrame')() for peak in peaks])
```

If we do this instead

```python
np.array([list(getattr(peak, 'getQSampleFrame')()) for peak in peaks])
```

it cuts the time to about 1 second.

Reply:
In that case we could use a trick we [introduced in the SANS backend](https://code.ornl.gov/sns-hfir-scse/sans/sans-backend/-/blob/next/drtsans/settings.py#L129). Running this in  my workstation:
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import time

LoadEmptyInstrument(InstrumentName='TOPAZ', OutputWorkspace='topaz')
CreatePeaksWorkspace(InstrumentWorkspace='topaz', NumberOfPeaks=100000, OutputType='LeanElasticPeak', OutputWorkspace='peaks')
peaks = mtd['peaks']

start = time.time()
np.array(peaks.column('QSample'))
print('Elapsed column = ', time.time() - start)

start = time.time()
np.array([getattr(peak, 'getQSampleFrame')() for peak in peaks])
print('Elapsed getattr = ', time.time() - start)

start = time.time()
np.array([list(getattr(peak, 'getQSampleFrame')()) for peak in peaks])
print('Elapsed list getattr = ', time.time() - start)

def unpack_pos(peak):
    v3d_vector = getattr(peak, 'getQSampleFrame')()
    return [v3d_vector.X(), v3d_vector.Y(), v3d_vector.Z()]
    
start = time.time()
np.array([unpack_pos(peak) for peak in peaks])
print('Elapsed unpack_pos = ', time.time() - start) 
```

Results in:

> LoadEmptyInstrument started
> LoadEmptyInstrument successful, Duration 6.34 seconds
> CreatePeaksWorkspace started
> CreatePeaksWorkspace successful, Duration 1.32 seconds
> Elapsed column =  9.059121370315552
> Elapsed getattr =  9.599326133728027
> Elapsed list getattr =  4.169772148132324
> Elapsed unpack_pos =  1.5399858951568604

Reply:
Yeah probably can't get much faster than that unless we move this to c++. This could be implemented as a method on the PeaksWorkspace, but I don't know if we need to do that at this point.